### PR TITLE
Numpy union1d

### PR DESCRIPTION
**Pull request by joschkabraun**
This is an implementation of [numpy.union1d](https://docs.scipy.org/doc/numpy/reference/generated/numpy.union1d.html) from issue #70 (and #2078). It follows the same structure as the original implementation with changes for Jax compatibility.

np.union1d is intrinsically hard to make jit-compatible since it depends on np.unique and its output shape depends on the values in the array. This problem will be resolved as soon as np.unique is jit-compaitible (#2760). 

The test check all one dimensional arrays and number dtypes which are not complex (as by #2760, unique does not work with complex dtypes). The test does not compile with jit since unique is not compatible with it.

### CODE REVIEW

File: jax/numpy/lax_numpy.py [hunk 1]
```diff
@@ -3174,6 +3174,14 @@ def unique(ar, return_index=False, return_inverse=False,
   raise NotImplementedError(
         "np.unique is not implemented for the axis argument")
 
+@_wraps(onp.union1d)
+def union1d(ar1, ar2):
+  if ar1.ndim == 0:
+    ar1 = array([ar1])
+  if ar2.ndim == 0:
+    ar2 = array([ar2])
```
Comment (jakevdp) [inline jax/numpy/lax_numpy.py:3182]: The numpy function accepts multi-dimensional inputs:
```python
>>> np.union1d(np.arange(9).reshape(3, 3), 10)                              
array([ 0,  1,  2,  3,  4,  5,  6,  7,  8, 10])
```
It seems like the easiest way to handle this, as well as the zero-dim case, would be to call ``ravel()`` on both inputs

File: tests/lax_numpy_test.py [hunk 2]
```diff
@@ -1128,6 +1128,23 @@ def testUnique(self, shape, dtype, return_index, return_inverse, return_counts,
...
     self._CheckAgainstNumpy(onp_fun, jnp_fun, args_maker, check_dtypes=True)
 
+  @parameterized.named_parameters(jtu.cases_from_list(
+      {"testcase_name": "_{}_{}".format(
+       jtu.format_shape_dtype_string(shape1, dtype1),
+       jtu.format_shape_dtype_string(shape2, dtype2)),
+       "shape1": shape1, "dtype1": dtype1, "shape2": shape2, "dtype2": dtype2,
+       "rng_factory": jtu.rand_default}
+      for dtype1 in [s for s in float_dtypes or int_dtypes if s != jnp.bfloat16]
+      for dtype2 in [s for s in float_dtypes or int_dtypes if s != jnp.bfloat16]
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:1138]: ``float_dtypes or int_dtypes`` is probably not doing what is intended. You can use ``default_dtypes`` to get the union of `float_dtypes` and `int_dtypes`.

File: tests/lax_numpy_test.py [hunk 3]
```diff
@@ -1128,6 +1128,23 @@ def testUnique(self, shape, dtype, return_index, return_inverse, return_counts,
...
+      for shape1 in one_dim_array_shapes or [jtu.NUMPY_SCALAR_SHAPE]
+      for shape2 in one_dim_array_shapes or [jtu.NUMPY_SCALAR_SHAPE]))
+  def testUnion1d(self, shape1, dtype1, shape2, dtype2, rng_factory):
+    rng = rng_factory()
+    args_maker = lambda: [rng(shape1, dtype1), rng(shape2, dtype2)]
+    onp_fun = lambda ar1, ar2: onp.union1d(ar1, ar2)
+    jnp_fun = lambda ar1, ar2: jnp.union1d(ar1, ar2)
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:1145]: These lambdas are a no-op, so you can drop them and use ``onp.union1d`` and ``jnp.union1d`` directly.

File: tests/lax_numpy_test.py [hunk 4]
```diff
@@ -1128,6 +1128,23 @@ def testUnique(self, shape, dtype, return_index, return_inverse, return_counts,
...
+  @parameterized.named_parameters(jtu.cases_from_list(
+      {"testcase_name": "_{}_{}".format(
+       jtu.format_shape_dtype_string(shape1, dtype1),
+       jtu.format_shape_dtype_string(shape2, dtype2)),
+       "shape1": shape1, "dtype1": dtype1, "shape2": shape2, "dtype2": dtype2,
+       "rng_factory": jtu.rand_default}
+      for dtype1 in [s for s in float_dtypes or int_dtypes if s != jnp.bfloat16]
+      for dtype2 in [s for s in float_dtypes or int_dtypes if s != jnp.bfloat16]
+      for shape1 in one_dim_array_shapes or [jtu.NUMPY_SCALAR_SHAPE]
+      for shape2 in one_dim_array_shapes or [jtu.NUMPY_SCALAR_SHAPE]))
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:1140]: Consider also testing multi-dimensional array shapes, as these are supported by numpy's version.

**Comment by joschkabraun** [comment]
Okay, thank you for your feedback! I will incorporate them and finish it on the weekend!

**Comment by joschkabraun** [comment]
Incorporated above changes!

**Comment by joschkabraun** [comment]
The tests are running smooth on my local machine but I am getting errors here which seem to be outside of my code. Can you help me with them?

**Comment by jakevdp** [comment]
We are in the process of migrating from using ``import numpy as onp; import jax.numpy as np`` to using ``import numpy as np; import jax.numpy as jnp``, and this PR is a bit caught in the middle.

The best thing to do would be to rebase on master, update your use of ``onp`` so the tests pass locally, and then force push to this branch (let me know if you're not clear on how to do that).

**Comment by jakevdp** [comment]
Looks good - please also remove the two .orig files that were added along the way.

**Comment by joschkabraun** [comment]
Okay, this has happened now!

File: jax/numpy/lax_numpy.py [hunk 5]
```diff
@@ -3320,6 +3320,12 @@ def unique(ar, return_index=False, return_inverse=False,
   raise NotImplementedError(
         "np.unique is not implemented for the axis argument")
 
+@_wraps(np.union1d)
+def union1d(ar1, ar2):
+  ar1 = ravel(asarray(ar1))
+  ar2 = ravel(asarray(ar2))
```
Comment (jakevdp) [inline jax/numpy/lax_numpy.py:3326]: Remove `asarray` in both cases here; `ravel()` already handles all valid inputs.

Comment (joschkabraun) [reply jax/numpy/lax_numpy.py]: Okay!

**Review by jakevdp — APPROVED** [review]
Looks great, Thanks!

**Comment by jakevdp** [comment]
As with #2943, we'll need a rebase against master before we can merge this.

**Comment by jakevdp** [comment]
Hello - are you still interested in finishing this contribution?

**Comment by jakevdp** [comment]
Since there has been no action here for a while, I'm going to close this. Feel free to re-open if you'd like to re-visit this contribution! Thanks
